[DataLoader] Unzip/Zip IterDataPipe

ejguan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #52701 [DataLoader] ConcatIterDataPipe
* **#52641 [DataLoader] Unzip/Zip IterDataPipe**

`UnzipIterDataPipe` is 'blocking' version of Routed DataPipe.

Example for (image, label)
```py
import torch.utils.data.datapipes as dp
import torchvision.transforms

image_label_dp  # [(image, label), ...]
img_dp, label_dp = dp.iter.Unzip(image_label_dp)

transforms = torchvision.transforms.Compose([
    torchvision.transforms.ToTensor(),
    torchvision.transforms.Pad(1, fill=1, padding_mode='constant'),
])
img_dp = dp.iter.Transforms(img_dp, transforms=transforms)

def label_fn(label):
    torch.tensor(label)
label_dp = dp.iter.Map(label_dp, fn=label_fn)

processed_dp = dp.iter.Zip(img_dp, label_dp)
```

Differential Revision: [D26595613](https://our.internmc.facebook.com/intern/diff/D26595613)

ejguan committed:
[DataLoader] Unzip/Zip IterDataPipe

[ghstack-poisoned]

ejguan committed:
Update on "[DataLoader] Unzip/Zip IterDataPipe"


Differential Revision: [D26595613](https://our.internmc.facebook.com/intern/diff/D26595613)

[ghstack-poisoned]

ejguan committed:
Update on "[DataLoader] Unzip/Zip IterDataPipe"


Differential Revision: [D26595613](https://our.internmc.facebook.com/intern/diff/D26595613)

[ghstack-poisoned]

VitalyFedyunin commented on torch/utils/data/datapipes/iter/combining.py:
```diff
@@ -0,0 +1,183 @@
+        source_dp = super().__new__(cls)
+        it = iter(datapipe)
+        try:
+            data = next(it)
```
It is not the best idea to start iterating items before completion of graph. I assume you want to type check on construction, but please move this check to runtime.

ejguan replied:
You are right, it makes sense not to start iteration at the construction time.
But, it raises another problem for type checking. I am wondering do you have any suggestion for all type checking at the construction time, if we do want this feature, we may need to force each DataPipe to specify it's own data type. Otherwise, most of type checking needs to happen at the runtime.

ejguan replied:
However, for `Unzip`, we need to know the number of elements to split the input DataPipe into pieces in the construction time. So, we have to fetch one data from input. But, we wont't need to run this duplicate iteration if I pass the first data and iterator to __init__.
I will update the PR and let you know.

ejguan committed:
Update on "[DataLoader] Unzip/Zip IterDataPipe"


`UnzipIterDataPipe` is 'blocking' version of Routed DataPipe.

Example for (image, label)
```py
import torch.utils.data.datapipes as dp
import torchvision.transforms

image_label_dp  # [(image, label), ...]
img_dp, label_dp = dp.iter.Unzip(image_label_dp)

transforms = torchvision.transforms.Compose([
    torchvision.transforms.ToTensor(),
    torchvision.transforms.Pad(1, fill=1, padding_mode='constant'),
])
img_dp = dp.iter.Transforms(img_dp, transforms=transforms)

def label_fn(label):
    torch.tensor(label)
label_dp = dp.iter.Map(label_dp, fn=label_fn)

processed_dp = dp.iter.Zip(img_dp, label_dp)
```

Differential Revision: [D26595613](https://our.internmc.facebook.com/intern/diff/D26595613)

[ghstack-poisoned]

ejguan commented on torch/utils/data/datapipes/iter/combining.py:
```diff
@@ -0,0 +1,188 @@
+        source_dp = super().__new__(cls)
+        # TODO(Erjia): Valid type inference mechanism without fetching data
+        #              at construction time
+        it = iter(datapipe)
+        data = next(it)
+        num_splits = len(data)
+        source_dp.__init__(datapipe, num_splits, it, data)
```
I have to get one data at the construction time to get the number of elements from the source DataPipe.
As @VitalyFedyunin suggested, it will ruin our intention to have lazy_init for the pipeline. In order to get rid of data fetching at construction time, we have to have a mechanism to infer or get the data type for all DataPipes.

ejguan commented:
Split the stack of PRs into multiple separate PRs in parallel.
